enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,238 @@
+`oauth.openshift.io/idp.<idpname>: "synced"`
+where `<idpname>` marks the identity provider name from the configuration. It should
+also get an `oauth.openshift.io/generated: true` if it hasn't previously existed.
+
+#### oauth-server
+
+When the OIDC provider has group synchronization configured, the oauth-server
+must attempt to retrieve the group claim (or claims) from the `id_token` it
+obtains from the provider after a successful authentication, and should do
+the same for the `id_token` retrieved from the OIDC's UserInfo endpoint
+should the OIDC be exposing this endpoint in its OIDC discovery data. The
+UserInfo response must be authoritative in that case.
+
+Once the server has the user's groups information, it will look up the
+OpenShift groups the user appears in, and it will either add to/remove from/create
+the necessary groups for the user. The user should only be removed from the groups
+that were created for that specific identity provider (as per the annotation
+from the [API section](#api)). If such a group ends up being empty, it should
+be removed if it contains the `oauth.openshift.io/generated: true` annotation.
+
+When a group that would otherwise be created on sync already exists, the
+oauth-server should only add the user among the group members and add an
+`oauth.openshift.io/idp.<idpname>` annotation to mark users of that IdP appear
+in that group.
+
+This implementation is not solving an issue when a user can be sourced from
+multiple identity providers when `mappingMethod: add` is configured for these
+providers. Such a user would get group memberships in all the groups from
+all matching identity providers. An alternative implementation will be
+discussed in the [Alternatives section](#alternatives).
+
+### Risks and Mitigations
+
+Given that the group synchronization only happens on logins, there is a time
+window during which the synchronization of groups between the identity provider
+and OpenShift is not guaranteed. This would typically be an issue when someone's
+group membership gets revoked.
+
+In order to force OpenShift group membership synchronization on group membership
+revocation on the identity provider side, one might:
+
+1. revoke all the OAuthAccessTokens for a given user by removing them, leveraging
+   a `userName` field selector, thus forcing the user to relogin:
+
+```bash
+       oc delete oauthaccesstoken --field-selector=userName="<username>"
+```
+
+2. remove the user from the OpenShift group manually
+
+Users can also be added to groups in the IdP, in which case they are more
+likely to willingly cause a resync by `oc login` themselves.
+
+## Design Details
+
+### Open Questions
+
+1. Do we care about mapping the groups to their source identity provider per
+   session (i.e. a user is only a member of the groups sourced from the IdP
+   that they logged in to + any OpenShift, non-synced groups)?
+
+### Test Plan
+
+The authentication operator already has a way to configure Keycloak and
```

Comment:
What are the 3rd party IdPs we'll use for testing? Keycloak? Google? Maybe something else?

Reply:
We can use anything that provides a container and has a REST API to work with users and groups. I wouldn't use more than a single IdP though, writing vendor-specific integration tests starts being time-consuming.

Reply:
Ok, this works for me. But at some point in time, we'll probably need to mention at least one "certified" IdP in the documentation (or show some example, how to use this feature). It would be perfect if that example, and what we test automatically were the same things.

Reply:
We may need to clarify what we expect of the OIDC provider to be capable of doing but this should generally work, without any IdP being put as "this product really works the best".

Reply:
Ok, so if we do not plan to certify any of the 3rd party IdP, we might need to implement a flexible mechanism for obtaining groups from the User Info Endpoint. In case of Keycloak, we know the structure (here's [an example](https://gist.github.com/slaskawi/647d903186f7af4c2a390ce9cd620e3d)) but I'm not sure we can assume the same layout for other IdPs. What do you think?

Reply:
Clarified off-list - this is where `OpenIDClaims` comes into play.